File: src/commands/hook-prompt.ts
import { Command } from 'commander';
import { discoverProject, openDatabase } from '../core/project.js';
import { getMessages, getMessagesWithMention } from '../db/queries.js';

interface HookOutput {
  additionalContext?: string;
}

export function hookPromptCommand(): Command {
  return new Command('hook-prompt')
    .description('UserPromptSubmit hook handler (internal)')
    .action(() => {
      const output: HookOutput = {};

      try {
        const agentId = process.env.MM_AGENT_ID;

        // If no agent ID, stay silent
        if (!agentId) {
          console.log(JSON.stringify(output));
          return;
        }

        // Try to discover beads project
        let project;
        try {
          project = discoverProject();
        } catch {
          // Not in a beads project - stay silent
          console.log(JSON.stringify(output));
          return;
        }

        const db = openDatabase(project);

        const roomLimit = 5;
        const mentionsLimit = 3;

        const roomMessages = getMessages(db, { limit: roomLimit });
        const agentBase = agentId.includes('.')
          ? agentId.substring(0, agentId.lastIndexOf('.'))
          : agentId;
        const mentionMessages = getMessagesWithMention(db, agentBase, { limit: mentionsLimit + roomLimit });

        // Filter out mentions already in room
        const roomIds = new Set(roomMessages.map(m => m.id));
        const filteredMentions = mentionMessages.filter(m => !roomIds.has(m.id)).slice(0, mentionsLimit);

        // Only inject if there's something to show
        if (roomMessages.length === 0 && filteredMentions.length === 0) {
          db.close();
          console.log(JSON.stringify(output));
          return;
        }

        let context = `[mm ${agentId}] `;

        // Compact format for per-prompt injection
        const parts: string[] = [];

        if (roomMessages.length > 0) {
          const lastMsg = roomMessages[roomMessages.length - 1];
          parts.push(`Room[${roomMessages.length}]: latest [${lastMsg.id}] ${lastMsg.from_agent}`);
        }

        if (filteredMentions.length > 0) {
          parts.push(`@mentions[${filteredMentions.length}]`);
          // Show the mentions briefly
          for (const msg of filteredMentions.slice(0, 2)) {
            const truncated = msg.body.length > 60 ? msg.body.slice(0, 60) + '...' : msg.body;
            parts.push(`  [${msg.id}] ${msg.from_agent}: ${truncated}`);
          }
        }

        context += parts.join(' | ');
        context += ` (mm get ${agentId} for full view)`;

        output.additionalContext = context;
        db.close();
      } catch {
        // On error, stay silent
      }

      console.log(JSON.stringify(output));
    });
}

File: internal/command/hook_prompt.go
package command

import (
	"fmt"
	"os"
	"strings"

	"github.com/adamavenir/mini-msg/internal/core"
	"github.com/adamavenir/mini-msg/internal/db"
	"github.com/adamavenir/mini-msg/internal/types"
	"github.com/spf13/cobra"
)

// NewHookPromptCmd handles Claude UserPromptSubmit hooks.
func NewHookPromptCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "hook-prompt",
		Short: "UserPromptSubmit hook handler (internal)",
		RunE: func(cmd *cobra.Command, args []string) error {
			output := hookOutput{}

			agentID := os.Getenv("MM_AGENT_ID")
			if agentID == "" {
				return writeHookOutput(cmd, output)
			}

			projectPath := os.Getenv("CLAUDE_PROJECT_DIR")
			project, err := core.DiscoverProject(projectPath)
			if err != nil {
				return writeHookOutput(cmd, output)
			}

			dbConn, err := db.OpenDatabase(project)
			if err != nil {
				return writeHookOutput(cmd, output)
			}
			defer dbConn.Close()
			if err := db.InitSchema(dbConn); err != nil {
				return writeHookOutput(cmd, output)
			}

			roomMessages, mentionMessages, _ := fetchHookMessages(dbConn, agentID, 5, 3)
			if len(roomMessages) == 0 && len(mentionMessages) == 0 {
				return writeHookOutput(cmd, output)
			}

			output.AdditionalContext = buildHookPromptContext(agentID, roomMessages, mentionMessages)
			return writeHookOutput(cmd, output)
		},
	}

	return cmd
}

func buildHookPromptContext(agentID string, roomMessages, mentionMessages []types.Message) string {
	var parts []string

	if len(roomMessages) > 0 {
		last := roomMessages[len(roomMessages)-1]
		parts = append(parts, fmt.Sprintf("Room[%d]: latest [%s] %s", len(roomMessages), last.ID, last.FromAgent))
	}

	if len(mentionMessages) > 0 {
		parts = append(parts, fmt.Sprintf("@mentions[%d]", len(mentionMessages)))
		limit := len(mentionMessages)
		if limit > 2 {
			limit = 2
		}
		for i := 0; i < limit; i++ {
			msg := mentionMessages[i]
			preview := msg.Body
			if len(preview) > 60 {
				preview = preview[:60] + "..."
			}
			parts = append(parts, fmt.Sprintf("  [%s] %s: %s", msg.ID, msg.FromAgent, preview))
		}
	}

	context := fmt.Sprintf("[mm %s] %s (mm get %s for full view)", agentID, strings.Join(parts, " | "), agentID)
	return context
}
